core/trino-main/src/test/java/io/trino/operator/TestDeduplicationExchangeClientBuffer.java

```diff
@@ -0,0 +1,443 @@
+    }
+
+    private void testPollPages(Multimap<TaskId, SerializedPage> pages, Map<TaskId, RuntimeException> failures, List<String> expectedValues)
+    {
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            for (TaskId taskId : Sets.union(pages.keySet(), failures.keySet())) {
+                buffer.addTask(taskId);
+            }
+            for (Map.Entry<TaskId, SerializedPage> page : pages.entries()) {
+                buffer.addPages(page.getKey(), ImmutableList.of(page.getValue()));
+            }
+            for (Map.Entry<TaskId, RuntimeException> failure : failures.entrySet()) {
+                buffer.taskFailed(failure.getKey(), failure.getValue());
+            }
+            for (TaskId taskId : Sets.difference(pages.keySet(), failures.keySet())) {
+                buffer.taskFinished(taskId);
+            }
+            buffer.noMoreTasks();
+            assertContainsPages(buffer, expectedValues);
+        }
+    }
+
+    private static void assertContainsPages(ExchangeClientBuffer buffer, List<String> values)
+    {
+        List<String> actual = new ArrayList<>();
+        while (true) {
+            SerializedPage page = buffer.pollPage();
+            if (page == null) {
+                break;
+            }
+            actual.add(page.getSlice().toStringUtf8());
+        }
+        assertTrue(buffer.isFinished());
+        assertThat(actual).containsExactlyInAnyOrderElementsOf(values);
+    }
+
+    @Test
+    public void testRemovePagesForPreviousAttempts()
+    {
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            assertEquals(buffer.getRetainedSizeInBytes(), 0);
+
+            TaskId task1 = createTaskId(0, 0);
+            TaskId task2 = createTaskId(1, 0);
+            TaskId task3 = createTaskId(0, 1);
+
+            SerializedPage page1 = createPage("textofrandomlength");
+            SerializedPage page2 = createPage("textwithdifferentlength");
+            SerializedPage page3 = createPage("smalltext");
+
+            buffer.addTask(task1);
+            buffer.addPages(task1, ImmutableList.of(page1));
+            buffer.addTask(task2);
+            buffer.addPages(task2, ImmutableList.of(page2));
+
+            assertEquals(buffer.getRetainedSizeInBytes(), page1.getRetainedSizeInBytes() + page2.getRetainedSizeInBytes());
+
+            buffer.addTask(task3);
+            assertEquals(buffer.getRetainedSizeInBytes(), 0);
+
+            buffer.addPages(task3, ImmutableList.of(page3));
+            assertEquals(buffer.getRetainedSizeInBytes(), page3.getRetainedSizeInBytes());
+        }
+    }
+
+    @Test
+    public void testBufferOverflow()
+    {
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), DataSize.of(100, BYTE), RetryPolicy.QUERY)) {
```

Comment:
it is not obvious from code flow that `page1.getRetainedSizeInBytes + page1.getRetainedSizeInBytes > 100`.
Maybe start the test like that:
```
            SerializedPage page1 = createPage("1234");
            SerializedPage page2 = createPage("123456789");

            assertThat(page1.getRetainedSizeInBytes()).isLessThanOrEqualTo(100);
            assertThat(page1.getRetainedSizeInBytes() + page2.getRetainedSizeInBytes()).isGreaterThan(100);
```

to make it more readable what is happening.